massa-execution/src/vm.rs

```diff
@@ -174,6 +177,63 @@ impl VM {
         context.ledger_step.caused_changes.clone()
     }
 
+    /// Run code in read-only mode
+    pub(crate) fn run_read_only(
+        &self,
+        slot: Slot,
+        max_gas: u64,
+        simulated_gas_price: Amount,
+        bytecode: Vec<u8>,
+        address: Option<Address>,
+        result_sender: oneshot::Sender<ExecuteReadOnlyResponse>,
+    ) {
+        let saved_changes = {
+            let mut context = self.execution_context.lock().unwrap();
+
+            // Set the call stack, using the provided address, or a random one.
+            let address = address.unwrap_or_else(|| {
+                let private_key = generate_random_private_key();
+                let public_key = derive_public_key(&private_key);
+                Address::from_public_key(&public_key)
+            });
+            context.call_stack = vec![address].into();
+
+            // Set the max gas.
+            context.max_gas = max_gas;
+
+            // Set the simulated gas price.
+            context.gas_price = simulated_gas_price;
+
+            // Return a clone of the current state, to revert to it after execution.
+            context.ledger_step.caused_changes.clone()
+        };
```

Comment:
It looks like something more generic, can we find a way to use the prepare_context or extract a function?

Reply:
I don't think its' worth it in this PR. If you see a clear alternative feels free to open an issue/pr and propose the change.

Reply:
we don't need to save/ pop changes at all actually :) just reset the context, run and that's it (just don't apply local_changes to the ledger at the end). The save/pop is there in run_step_internal because we need to keep some changes but not others, and we always apply at least some changes to the final ledger. Here it's not the case, so we can simplify :)